core/src/main/java/com/linecorp/armeria/common/stream/AbstractStreamMessageDuplicator.java

```diff
@@ -0,0 +1,601 @@
+                ChildStreamMessage.class, DownstreamSubscription.class, "subscription");
+
+        private StreamMessageProcessor processor;
+        @SuppressWarnings("unused")
+        private volatile DownstreamSubscription subscription;
+
+        private CompletableFuture<Void> closeFuture = new CompletableFuture<>();
+
+        ChildStreamMessage(StreamMessageProcessor<? extends T> processor) {
+            this.processor = processor;
+        }
+
+        @Override
+        public boolean isOpen() {
+            return !closeFuture.isDone();
+        }
+
+        @Override
+        public boolean isEmpty() {
+            return !isOpen() && processor.upstreamOffset == 0;
+        }
+
+        @Override
+        public CompletableFuture<Void> closeFuture() {
+            return closeFuture;
+        }
+
+        @Override
+        public void subscribe(Subscriber<? super T> subscriber) {
+            requireNonNull(subscriber, "subscriber");
+            subscribe(subscriber, false);
+        }
+
+        @Override
+        public void subscribe(Subscriber<? super T> subscriber, boolean withPooledObjects) {
+            requireNonNull(subscriber, "subscriber");
+            subscribe0(subscriber, null, withPooledObjects);
+        }
+
+        @Override
+        public void subscribe(Subscriber<? super T> subscriber, Executor executor) {
+            requireNonNull(subscriber, "subscriber");
+            requireNonNull(executor, "executor");
+            subscribe(subscriber, executor, false);
+        }
+
+        @Override
+        public void subscribe(Subscriber<? super T> subscriber, Executor executor,
+                              boolean withPooledObjects) {
+            requireNonNull(subscriber, "subscriber");
+            requireNonNull(executor, "executor");
+            subscribe0(subscriber, executor, withPooledObjects);
+        }
+
+        private void subscribe0(Subscriber<? super T> subscriber, Executor executor,
+                                boolean withPooledObjects) {
+            final DownstreamSubscription subscription = new DownstreamSubscription(
+                    this, subscriber, processor, executor, withPooledObjects);
+            if (!subscriptionUpdater.compareAndSet(this, null, subscription)) {
+                throw new IllegalStateException(
+                        "Subscribed by other subscriber already: " + this.subscription.subscriber());
+            }
+            processor.subscribe(subscription);
+        }
+
+        @Override
+        public void abort() {
+            final DownstreamSubscription currentSubscription = subscription;
+            if (currentSubscription != null) {
+                currentSubscription.cancel();
+                return;
+            }
+
+            final DownstreamSubscription newSubscription = new DownstreamSubscription(
+                    this, AbortingSubscriber.INSTANCE, null, null, false);
```

Comment:
The subscription on the line 483 is the subscription instance which is made at the line 461, so it is ok to call the cancel because it already have processor.